source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc

```diff
@@ -71,58 +118,258 @@ absl::optional<Arn> Filter::calculateRouteArn() {
     return Http::FilterHeadersStatus::Continue;
   }
 
-  auto route_arn = calculateRouteArn();
-  if (route_arn.has_value()) {
-    auto cluster_info_ptr = decoder_callbacks_->clusterInfo();
-    ASSERT(cluster_info_ptr);
-    if (!isTargetClusterLambdaGateway(*cluster_info_ptr)) {
-      skip_ = true;
-      return Http::FilterHeadersStatus::Continue;
-    }
-    arn_.swap(route_arn);
-  } else {
+  const auto err = resolveSettings();
+
+  if (!err.empty()) {
+    skip_ = true;
+    decoder_callbacks_->sendLocalReply(Http::Code::BadRequest, err, nullptr /*modify_headers*/,
+                                       absl::nullopt /*grpc_status*/, "" /*details*/);
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  if (!arn_) {
     arn_ = parseArn(settings_.arn());
     if (!arn_.has_value()) {
-      ENVOY_LOG(error, "Unable to parse Lambda ARN {}.", settings_.arn());
+      ENVOY_LOG(error, "Failed to parse Lambda ARN {}.", settings_.arn());
       skip_ = true;
-      return Http::FilterHeadersStatus::Continue;
+      decoder_callbacks_->sendLocalReply(Http::Code::BadRequest, "Invalid AWS Lambda ARN",
+                                         nullptr /*modify_headers*/, absl::nullopt /*grpc_status*/,
+                                         "" /*details*/);
+      return Http::FilterHeadersStatus::StopIteration;
     }
   }
 
-  if (end_stream) {
-    setHeaders(headers, arn_->functionName());
+  if (!end_stream) {
+    request_headers_ = &headers;
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  if (payload_passthrough_) {
+    setLambdaHeaders(headers, arn_->functionName());
     sigv4_signer_->sign(headers);
     return Http::FilterHeadersStatus::Continue;
   }
 
-  headers_ = &headers;
+  Buffer::OwnedImpl json_buf;
+  jsonizeRequest(headers, nullptr, json_buf);
+  // We must call setLambdaHeaders *after* the JSON transformation of the request. That way we
+  // reflect the actual incoming request headers instead of the overwritten ones.
+  setLambdaHeaders(headers, arn_->functionName());
+  headers.setContentLength(json_buf.length());
+  headers.setReferenceContentType(content_type_json);
+  auto& hashing_util = Envoy::Common::Crypto::UtilitySingleton::get();
+  const auto hash = Hex::encode(hashing_util.getSha256Digest(json_buf));
+  sigv4_signer_->sign(headers, hash);
+  decoder_callbacks_->addDecodedData(json_buf, false);
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap& headers, bool end_stream) {
+  if (skip_ || end_stream) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // Check for errors returned by Lambda.
+  // If we detect an error, we skip the encodeData step to hand the error back to the user as is.
+  // Errors can be in the form of HTTP status code or x-amz-function-error header
+  if (auto http_status_header = headers.Status()) {
```

Comment:
Use getResponseStatus here. There should always be a status so you don't need an if statement.